private/pkg/manifest/manifest.go

```diff
@@ -136,24 +178,26 @@ func (m *Manifest) AddEntry(path string, digest Digest) error {
 	return nil
 }
 
-// Paths returns all unique paths in the manifest, order not guaranteed. If you want to iterate the
-// paths and their digests, consider using `Range` instead.
+// Paths returns all unique paths in the manifest by insertion order.
```

Comment:
I think not guaranteeing order is actually better, so the callers don't assume we have the blobs in the manifest ordered somehow (which we didn't before this PR). I guess I don't see what this gets us.

Reply:
This is the same argument used in the `protojson` package as to why they introduce random white space in their marshalling of Protobuf messages to JSON, and it causes similar issues downstream.

In this case, it caused `Walk` to not guarantee iteration order, which all the other `Bucket` implementations do, as the implementer of the manifest `Bucket` did not take into account the non-deterministic iteration order. There was also documentation saying "use `Range` if you want your ordering to be deterministic", however range was no deterministic either.

Even more to the point, one of the majority use cases of `Manifests`, namely `MarshalText`, was already guaranteeing ordering implicitly via `sort.Strings`.

I've seen examples of code introducing the chaos monkey principle in various parts of the codebase (such as `protojson`), and more often than not, it causes bugs and doesn't prevent other bugs. In this example, it caused an effective bug in `Walk`, which could cause functions such as file listing to be non-deterministic.

Reply:
> it caused Walk to not guarantee iteration order, which all the other Bucket implementations do

I didn't know that, probably we should add it to the `Walk` interface func docs, so every other implementer knows this rule.

> There was also documentation saying "use Range if you want your ordering to be deterministic", however range was no deterministic either.

[The previous Range documentation](https://github.com/bufbuild/buf/pull/2284/files#diff-536233f971a641a3a3fc1c6fb13958674f984e202f8dc0cf7e07c0f28ff741eaL166-L168) explicitly said the order was not guaranteed.

> Even more to the point, one of the majority use cases of Manifests, namely MarshalText, was already guaranteeing ordering implicitly via sort.Strings.

Correct, because that was the only place in which we were interested in sorting the paths, when returning the manifest as a blob, so the whole module's digest would be deterministic. Order adds overhead calculation, even though is not required anywhere else in the implementation.

> In this example, it caused an effective bug in Walk, which could cause functions such as file listing to be non-deterministic.

You're right, I could see a scenario in which the Walk caller expects the files to be walked in a deterministic order, even though that is not explicitly stated in the docs, which is why we didn't consider it as a bug. I still don't think it is, but probably it could be seen as a subpar experience using the pkg. Tbh, in our usages of walking this manifest bucket, we hadn't seen the need of a deterministic order.

Reply:
I think there's a bit of philosophy on this, but I tend to default to making iteration deterministic if possible, and it's for this reason - engineers will just assume it is unless you tell them otherwise. You're totally right that it's not documented in `Walk`,  but had `Walk` not been done deterministically, I guarantee you there'd be downstream bugs in users of it.